Feature/mosmix forecasts

IvannikovG opened the pull request:
Might cause a bit trouble with measured values, because this feature reuses a lot of functionality from there. Tests are very basic too

MoritzStueber committed:
Add OAS for how to get forecasts at the location of a weather station

MoritzStueber committed:
Register handler for getting forecasts at a station's location (TBD)

Commit:
Implement first version (only for mosmix model and only json)

Commit:
Remove dropping NaNs from data

Commit:
Add csv and json rendering for mosmix model

Commit:
Restructure libs with helper functions

Commit:
Add full implementation for mosmix model and helper to dispatch for cosmo-d2 and mosmix behaviour

Commit:
Prepare functions to load data from cosmo-d2 directories, olds and new

Commit:
Add implementation for cosmo-d2 yet with no csv representation

Commit:
Add implementation for cosmo-d2 csv and unit conversion bug fix for cosmo-d2

Commit:
Add basic API testing for forecast endpoint

MoritzStueber commented on handlers/data_access.js:
```diff
@@ -13,10 +13,22 @@ const gf = require('../lib/grib_functions')
 // Instantiate logger
 const log = require('../lib/logger.js')
 
+const now = moment()
```
Careful, this code only gets executed when the module is first loaded, which could be months before a request arrives! Your attempt to reduce code duplication -- although honourable -- backfired :fire: :arrow_right: move back to request handlers

IvannikovG replied:
Moved

MoritzStueber commented on handlers/data_access.js:
```diff
@@ -369,49 +381,25 @@ function getMeasuredValues (WEATHER_DATA_BASE_PATH, voisConfigs) {
+      : parseInt(defaultStartTimestamp)
+    const endTimestamp = parseInt(req.query.to)
+      ? parseInt(req.query.from)
+      : parseInt(defaultEndTimestamp)
```
Ternary operators, cool :sunglasses:

MoritzStueber commented on lib/csv.js:
```diff
@@ -1,20 +1,64 @@
-// SPDX-FileCopyrightText: 2018 UdS AES <https://www.uni-saarland.de/lehrstuhl/frey.html>
-// SPDX-License-Identifier: MIT
+// dwd-csv-helper
+//
+// Copyright 2018 The dwd-csv-helper Developers. See the LICENSE file at
+// the top-level directory of this distribution and at
+// https://github.com/UdSAES/dwd-csv-helper/LICENSE
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
+// REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
+// AND FITNESS.IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
+// INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
+// LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
+// OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
+// PERFORMANCE OF THIS SOFTWARE.
+//
+// dwd-csv-helper may be freely used and distributed under the ISC license
```
The other way round, please -- we want to use the short form suggested by SPDX instead of the verbose, end- and mindless copy-paste of license texts everywhere.

I'm guessing that changing the license from ISC to MIT is theoretically more complex than changing three characters, but I'm certain that these two licenses share the same intent and I'm therefore suggesting to use MIT throughout `dwd_data_access`

MoritzStueber commented on lib/forecast_utils.js:
```diff
@@ -0,0 +1,379 @@
+  return csvLabels + csvStrings.join('\n')
+}
+
+function dropTimeseriesDataNotOfInteresWithParameter (
```
Typo in function name, please refactor

MoritzStueber replied:
Nevermind, function is gone

MoritzStueber commented on lib/forecast_utils.js:
```diff
@@ -0,0 +1,379 @@
+const log = bunyan.createLogger({
+  name: 'dwd-csv-handler',
+  serializers: bunyan.stdSerializers
+})
```
use the one instantiated in `lib/logger.js` instead

IvannikovG replied:
forecast utils was separated in a way that id does not include logger anymore. just pure functions

MoritzStueber commented on lib/forecast_utils.js:
```diff
@@ -0,0 +1,379 @@
+      jsonRenderer: renderJSONCosmoMosmix,
+      csvRenderer: renderForecastCSVCosmoMosmix
+    }
+  } else {
+    return {}
+  }
```
Reaching this branch would indicate some mistake, thus the appropriate action would be to throw an error. Returning an empty dictionary is problematic because you overload this empty dictionary with the meaning "this is an error", which should instead be communicated by throwing properly.

MoritzStueber replied:
Probably not that important because invalid model identifiers supplied by the user get caught before they ever reach this function, but still...

IvannikovG replied:
I think it will 100% catch incorrect model in the step when we analyse query string, so this is even optional step. My opinion - let us keep it like it is

MoritzStueber commented on handlers/data_access.js:
```diff
@@ -475,8 +463,124 @@ function getMeasuredValues (WEATHER_DATA_BASE_PATH, voisConfigs) {
+      })
+      req.log.warn(
+        { res: res },
+        'received request for REPORT for unconfigured MODEL-RUN'
```
`detail` and log message should be the same; `REPORT` is wrong; `model-run` is not unconfigured, but faulty; capitalization unneccessary

:arrow_right: change to `User-supplied value for parameter 'model-run' does not match available choices for model ${model}`

IvannikovG replied:
Makes sense

MoritzStueber commented on handlers/data_access.js:
```diff
@@ -475,8 +463,124 @@ function getMeasuredValues (WEATHER_DATA_BASE_PATH, voisConfigs) {
+      return
+    }
+
+    log.debug('reading BEOB data from disk...')
```
wrong, `BEOB` refers to measured values only

IvannikovG replied:
renamed to 'reading timeseries data from disk...'

MoritzStueber commented on lib/csv.js:
```diff
@@ -1,20 +1,64 @@
-function deriveCsvFilePath (csvBasePath, dayTimestamp, stationId) {
-  const dayDateTimeString = moment.utc(dayTimestamp).format('YYYYMMDD')
-  const fileName = stationId + '-BEOB.csv'
+const bunyan = require('bunyan')
+const log = bunyan.createLogger({
+  name: 'dwd-csv-handler',
+  serializers: bunyan.stdSerializers
+})
```
use the one in `lib/logger.js`

IvannikovG replied:
fixed it

MoritzStueber commented on lib/csv.js:
```diff
@@ -1,20 +1,64 @@
+})
 
-  return path.join(csvBasePath, dayDateTimeString, fileName)
+function deriveCsvFilePath (csvBasePath, type, timestamp, stationId) {
```
:warning: I will stop reviewing this file in this PR since it is currently not used by any other code.

MoritzStueber commented on lib/measured_values_utils.js:
```diff
@@ -4,7 +4,8 @@
 const _ = require('lodash')
-const { convertUnit } = require('../lib/unit_conversion.js')
+const gu = require('../lib/general_utils.js')
+
```
This probably breaks loading measurement data entirely... Commit more carefully and commit more often :wink:

MoritzStueber replied:
Ok, I see -- doesn't break the code because the function is no longer called. Nonetheless, this means that you now have dead code lying around (not good).

I will take care of this while restructuring.

MoritzStueber commented on lib/forecast_utils.js:
```diff
@@ -0,0 +1,379 @@
+  return result
+}
+
+function checkIfValidModelRun (allowed, modelRun) {
+  return allowed.includes(modelRun)
+}
```
Programmers often name such test functions that return `true` or `false` `isSomething`, e.g. `isValidModelRun` :arrow_right: tests can then be read more naturally, e.g. ` if (!isValidModelRun())` reads "if not is valid model run" as compared to "if not check if valid model run" -- subtle difference, but I think it makes sense to rename

IvannikovG replied:
Renamed

MoritzStueber commented on handlers/data_access.js:
```diff
@@ -475,8 +463,124 @@ function getMeasuredValues (WEATHER_DATA_BASE_PATH, voisConfigs) {
+    log.debug('rendering and sending response now')
+    res.format({
+      'application/json': async function () {
+        const localForecast = await config.jsonRenderer(
```
rename `localForecast` to `forecastRepresentation` because the latter is more precise -- also line 562, 566, 567

IvannikovG replied:
Renamed

MoritzStueber commented on handlers/data_access.js:
```diff
@@ -475,8 +463,124 @@ function getMeasuredValues (WEATHER_DATA_BASE_PATH, voisConfigs) {
+      )
+      return
+    }
+    const vois = gu.getVoisNamesFromQuery(req.query)
```
move down one line so that empty line separates code block for handling invalid user input from loading config for VOIs in case the input was correct

MoritzStueber commented on lib/forecast_utils.js:
```diff
@@ -0,0 +1,379 @@
+    moment
+      .utc(timestamp)
+      .format(formatString)
+      .slice(0, -2) + modelRun
```
will break reading the measured values (case `REPORT`) because the assumption on which `.slice(0, -2)` operates no longer holds

MoritzStueber replied:
resolve when integrating `dwd-csv-helper` into code base; discuss with me first

MoritzStueber commented on handlers/data_access.js:
```diff
@@ -369,49 +381,25 @@ function getMeasuredValues (WEATHER_DATA_BASE_PATH, voisConfigs) {
+      ? parseInt(req.query.from)
+      : parseInt(defaultStartTimestamp)
+    const endTimestamp = parseInt(req.query.to)
+      ? parseInt(req.query.from)
```
`req.query.to` !!

MoritzStueber replied:
fixed that

MoritzStueber commented on handlers/data_access.js:
```diff
@@ -475,8 +463,124 @@ function getMeasuredValues (WEATHER_DATA_BASE_PATH, voisConfigs) {
+    // const timeseriesDataArrayUnformatted = mvu.dropNaN(
+    //   mvu.dropTimeseriesDataNotOfInteresWithParameter(voiConfigs, timeseriesDataCollection, "mosmix")
+    // )
+    // log.trace({ timeseriesDataArrayUnformatted })
```
if it doesn't serve a purpose, it should not have been included in commit/PR :wink:

IvannikovG replied:
👍

MoritzStueber reviewed (changes requested):
Hi, I reviewed the entire pull request except `test/test_forecast_openapi_validation.js`.

Overall, I'm happy that you managed to pull this off -- it's been quite a task and I know from experience that understanding all the existing code is hard.

However, you introduced quite a bit of duplicated code throughout the files and the source files became quite hard to navigate. Therefore, I will restructure the code and reorganize files in order to hopefully increase ease of navigation.

Please review the comments on specific sections of code to get some feedback on your coding, but wait for my pull request before implementing any of the requested changes. I will start restructuring now, and aim to complete it by tomorrow (sunday) evening.

MoritzStueber committed:
Move helper functions for parsing info from requests to separate file

MoritzStueber committed:
Rename file containing code for importing lists of weather stations

MoritzStueber committed:
Rename files implementing utility functions for building station list

Also rename directory containing corresponding test data.

MoritzStueber committed:
Move code to read COSMO-D2 forecasts as timeseries to separate file

MoritzStueber committed:
Add code previously in separate package `dwd-grib-helper` to repo

No changes; as of commit 57f6cee692538bbc174bce0e0c9378e4d0fee3e7 on 
https://github.com/UdSAES/dwd-grib-helper; not integrated yet.

MoritzStueber committed:
Integrate code previously in package `dwd-grib-helper`; format/annotate

Also remove dependency on `dwd-grib-helper`.

MoritzStueber committed:
Change license of code previously in package `dwd-grib-helper` to MIT

Both ISC and MIT share the same intent, don't they?

MoritzStueber committed:
Delete code to find one specific .grib2.lz4-file which is no longer used

MoritzStueber committed:
Rename `lib/grib_functions.js` to `lib/cosmo-d2_functions.js`

... in order to use consistent file names.

MoritzStueber committed:
Move code for rendering timeseries as CSV to separate file

Also move corresponding unit tests to separate file and clean up 
duplicated code for rendering as CSV.

MoritzStueber committed:
Move code for rendering timeseries as JSON to separate file; simplify

MoritzStueber committed:
Remove duplicate log entry created for every request for whatever reason

MoritzStueber committed:
Remove duplicate function for unit conversion for array of timeseries

MoritzStueber committed:
Fix faulty rendering of timestamps

MoritzStueber committed:
Condense several functions for unit conversion on timeseries to one

Unit conversion for COSMO-D2 models already happens in 
`cd2f.loadBaseValue()`, therefore the function that does nothing but 
returning its second argument.

MoritzStueber committed:
Move code to identify MOSMIX-files to be read to separate file

MoritzStueber committed:
Remove old handler for accessing MOSMIX-forecast (now obsolete)

MoritzStueber committed:
Remove duplicated code for responding to requests with `406`

MoritzStueber committed:
Reuse code for parsing station ID from request URL instead of inlining

MoritzStueber committed:
Fix minuscule style issue

MoritzStueber committed:
Rename file containing handlers for station-specific resources

MoritzStueber committed:
Remove duplicated code from test definition

MoritzStueber committed:
Rename test files to enforce consistent naming scheme

MoritzStueber committed:
Fix faulty regular expression for quantity identifiers

MoritzStueber committed:
Comment out unused Chai plugin for validation against OAS

MoritzStueber committed:
Amend test harness to support forecasts at a specific weather station

Get rid of separate test file with lots of duplicated code; add two 
initial test cases for forecast at station.

MoritzStueber committed:
Include full link of request in test output for convenience/ debugging

MoritzStueber committed:
Identify API tests based on pattern of file name, not description string

MoritzStueber committed:
Use common code for disecting request

MoritzStueber committed:
Amend logging (of internal objects) for improved usability

MoritzStueber committed:
Fix bug causing empty timeseries by using start timestamp as end

MoritzStueber committed:
Only attempt unit conversion for non-empty collection of timeseries

MoritzStueber committed:
Fix crash upon rendering empty series as `text/csv`

MoritzStueber committed:
Change test definitions to access more diverse range of quantities

MoritzStueber committed:
Clean up comments/whitespaces (cosmetics only)

MoritzStueber committed:
Remove unused utility functions

MoritzStueber committed:
Identify unit tests by absence of `_api_` in file name

Replaces more brittle heuristics based on test name.

MoritzStueber commented:
> Please review the comments on specific sections of code to get some feedback on your coding, but wait for my pull request before implementing any of the requested changes. I will start restructuring now, and aim to complete it by tomorrow (sunday) evening.

Done: PR #6 

Some of my comments above will have resolved themselves by now, you'll see.

IvannikovG committed:
Merge pull request #6 from UdSAES/maintenance/restructuring

Merging restructuring commit

Commit:
Fixed small issues from pull request (renaming/ move moment js into handlers_
)

Commit:
replace old logger in grib2.js

Commit:
Fix failing json representation for cosmo-d2

Commit:
Fix csv representation for cosmo-d2

Commit:
Add explicit time limitations for timeseries

Commit:
Fix convertUnit implementation. Cosmo-d2 and mosmix timestamps are correctly limited from from to to

Commit:
Fix linter issues
